src/document_basic.c

```diff
@@ -214,28 +262,59 @@ void Document_Free(Document *doc) {
   }
 }
 
-int Redis_SaveDocument(RedisSearchCtx *ctx, Document *doc, int options, QueryError *status) {
-  RedisModuleKey *k =
-      RedisModule_OpenKey(ctx->redisCtx, doc->docKey, REDISMODULE_WRITE | REDISMODULE_READ);
-  if (k == NULL || (RedisModule_KeyType(k) != REDISMODULE_KEYTYPE_EMPTY &&
-                    RedisModule_KeyType(k) != REDISMODULE_KEYTYPE_HASH)) {
-    QueryError_SetError(status, QUERY_EREDISKEYTYPE, NULL);
-    if (k) {
-      RedisModule_CloseKey(k);
-    }
-    return REDISMODULE_ERR;
+static RedisModuleString **globalAddRSstrings = NULL;
+
+static void initGlobalAddStrings() {
+  globalAddRSstrings = rm_malloc(3 * sizeof(*globalAddRSstrings));
+  const char *Sscore = "__score";
+  const char *Slang = "__language";
+  const char *Spayload = "__payload";
+
+  globalAddRSstrings[0] = RedisModule_CreateString(NULL, Sscore, strlen(Sscore)); 
+  globalAddRSstrings[1] = RedisModule_CreateString(NULL, Slang, strlen(Slang)); 
+  globalAddRSstrings[2] = RedisModule_CreateString(NULL, Spayload, strlen(Spayload)); 
+}
+
+void freeGlobalAddStrings() {
+  if (globalAddRSstrings == NULL) return;
+
+  for (size_t i = 0; i < 3; ++i) {
+    RedisModule_FreeString(NULL, globalAddRSstrings[i]);
   }
-  if ((options & REDIS_SAVEDOC_NOCREATE) && RedisModule_KeyType(k) == REDISMODULE_KEYTYPE_EMPTY) {
-    RedisModule_CloseKey(k);
-    QueryError_SetError(status, QUERY_ENODOC, "Document does not exist");
-    return REDISMODULE_ERR;
+  rm_free(globalAddRSstrings);
+  globalAddRSstrings = NULL;
+}
+
+int Redis_SaveDocument(RedisSearchCtx *ctx, const AddDocumentOptions *opts, QueryError *status) {
+  if (!globalAddRSstrings) {
+    initGlobalAddStrings();
   }
 
-  for (int i = 0; i < doc->numFields; i++) {
-    RedisModule_HashSet(k, REDISMODULE_HASH_CFIELDS, doc->fields[i].name, doc->fields[i].text,
-                        NULL);
+  // create an array for key + all field/value + score/language/payload
+  arrayof(RedisModuleString*) arguments = array_new(RedisModuleString*, 1 + opts->numFieldElems + 6);
+
+  arguments = array_append(arguments, opts->keyStr);
+  arguments = array_ensure_append_n(arguments, opts->fieldsArray, opts->numFieldElems);
+
+  if (opts->score != 1.0) {
+    arguments = array_append(arguments, globalAddRSstrings[0]);
+    arguments = array_append(arguments, opts->scoreStr);
   }
-  RedisModule_CloseKey(k);
+
+  if (opts->languageStr) {
+    arguments = array_append(arguments, globalAddRSstrings[1]);
+    arguments = array_append(arguments, opts->languageStr);
+  }
+
+  if (opts->payload) {
+    arguments = array_append(arguments, globalAddRSstrings[2]);
+    arguments = array_append(arguments, opts->payload);
+  }
+
+  RedisModule_Call(ctx->redisCtx, "HSET", "!v", arguments, array_len(arguments));
```

Comment:
Notice that if you choose to replicate the hset command (as you do with `!`) you should not relicate ft.add command ...